### ISSUE TITLE
needless_collect false positives around borrowing

### ISSUE THREAD

**Issue by Manishearth** [rust-lang/rust-clippy#7526]
Lint name: `needless_collect`

I tried this code:

```rust
struct Foo(Vec<u8>);

impl Foo {
    fn get_iterator(&self) -> impl Iterator<Item = u8> {
        let items: Vec<_> = self.0.iter().filter(|x| **x > 0).copied().collect();
        items.into_iter()
    }
}

```

([playpen](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=843ee625c223765e9059fc627baa47bb))

I expected to see this happen: No lint, because removing the collect produces an iterator that borrows from `self`

Instead, this happened:

```
warning: avoid using `collect()` when not needed
 --> src/main.rs:5:72
  |
5 |         let items: Vec<_> = self.0.iter().filter(|x| **x > 0).copied().collect();
  |                                                                        ^^^^^^^
6 |         items.into_iter()
  |         ----------------- the iterator could be used here instead
  |
  = note: `#[warn(clippy::needless_collect)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_collect
help: use the original Iterator instead of collecting it and then producing a new one
  |
5 |         
6 |         self.0.iter().filter(|x| **x > 0).copied()
  |
```

### Meta

- `cargo clippy -V`: `clippy 0.1.54 (a178d03 2021-07-26)`
- `rustc -Vv`:
```
rustc 1.54.0 (a178d0322 2021-07-26)
binary: rustc
commit-hash: a178d0322ce20e33eac124758e837cbd80a6f633
commit-date: 2021-07-26
host: x86_64-unknown-linux-gnu
release: 1.54.0
LLVM version: 12.0.1rustc 1.54.0 (a178d0322 2021-07-26)
binary: rustc
commit-hash: a178d0322ce20e33eac124758e837cbd80a6f633
commit-date: 2021-07-26
host: x86_64-unknown-linux-gnu
release: 1.54.0
LLVM version: 12.0.1
```

**Comment by dswij** [comment]
I think this is closely related to #7512

**Comment by hellow554** [comment]
I mean, you're right, but shouldn't this be linted anyway and say, that your `impl Iterator` should have a lifetime bound to self, e.g. : `impl Iterator<Item = u8> + '_`? Not sure if that's possible though.

**Comment by Manishearth** [comment]
@hellow554 It depends on what's being borrowed. We _could_ do this, however:

 - What if a trait expects that signature?
 - What if a local is being borrowed here
 - What if the method is being used in a way that its caller needs that signature?

The first two can be handled, the last can't, though we can declare that we're okay with it.

**Comment by khuey** [comment]
Every needless_collect warning on my ~250kloc Rust codebase is a false positive because of lifetime issues like this.

**Comment by osa1** [comment]
In case you need real-world examples here's another false positive: https://github.com/osa1/tiny/blob/8f4428ee41b46152fd4c2f1abf883a2f1942ceaa/crates/libtiny_tui/src/tui.rs#L1056-L1058
